agent/engine/task_manager.go

```diff
@@ -423,6 +450,50 @@ func (mtask *managedTask) handleContainerChange(containerChange dockerContainerC
 		mtask.Arn, container.Name, mtask.GetDesiredStatus().String())
 }
 
+// handleResourceStateChange attempts to update resource's known status depending on
+// the current status and errors during transition
+func (mtask *managedTask) handleResourceStateChange(resChange resourceStateChange) {
+	// locate the resource
+	res := resChange.resource
+	if !mtask.isResourceFound(res) {
+		seelog.Errorf("Managed task [%s]: state error; invoked with another task's resource [%s]",
+			mtask.Arn, res.GetName())
+		return
+	}
+
+	status := resChange.nextState
+	err := resChange.err
+	currentKnownStatus := res.GetKnownStatus()
+
+	if status <= currentKnownStatus {
+		seelog.Infof("Managed task [%s]: redundant resource state change. %s to %s, but already %s",
+			mtask.Arn, res.GetName(), res.StatusString(status), res.StatusString(currentKnownStatus))
+		return
+	}
+
+	if err == nil {
+		res.SetKnownStatus(status)
+		mtask.engine.saver.Save()
+		return
+	}
+	seelog.Debugf("Managed task [%s]: error while transitioning resource %s to %s: %v",
+		mtask.Arn, res.GetName(), res.StatusString(status), err)
+	if status == res.SteadyState() {
+		seelog.Errorf("Managed task [%s]: error while creating resource %s, setting the task's desired status to STOPPED",
+			mtask.Arn, res.GetName())
+		mtask.SetDesiredStatus(api.TaskStopped)
+		mtask.engine.saver.Save()
+	}
+}
+
+func (mtask *managedTask) emitResourceChange(change resourceStateChange) {
```

Comment:
You will need to do something similar to [this](https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/task_manager.go#L894) during cleanup, and also close this channel after the clean up.

Reply:
done.